File: java/bst/DeleteNodeInBST.java
package bst;

public class DeleteNodeInBST {
    /* 首先是二叉树搜索没有疑问 主要的难点在于如何找到node后进行删除
     * 这里首先要明白我们return的是root 如果要在找到的那个block就把整个树改好太困难了 要保持parent的信息等等
     * 所以我们在搜索的时候用比如root.right = deleteNode(root.right, key); 这样直接用递归上来的root更新我们的right 会方便很多
     * 然后明白几种情况
     * 1.leaf 直接return null 代表不要了
     * 2. 有左右两边其中之一 我们就把child替换上来
     * 3. root有两个child 我们要找到右边子树的最小node 把它替换上来
     * 怎么把它替换上来了 当然找到它 然后把它的left更新成root left right更新成root right
     * 在此之前 我们需要从右子树中把它删掉 不然不就circle了 删掉就可以直接递归了 这能保证找到最后是一个leaf的
     */
    public TreeNode deleteNode(TreeNode root, int key) {
        if(root == null) return null;
        if(root.val == key) {
            // delete
            if(root.left == null && root.right == null) {
                // this node is leaf
                return null; //为什么return null就可以了呢 这是因为call的那边正在做update 我们不必在这里对树作调整了
            }else if (root.left == null && root.right != null) {
                return root.right;
            }else if(root.right == null && root.left != null) {
                return root.left;
            }else {
                // find the smallest node from the right sub tree
                TreeNode cur = root.right; // right must not be null
                while(cur.left != null) {
                    cur = cur.left;
                }
                root.right = deleteNode(root.right, cur.val);
                cur.left = root.left;
                cur.right = root.right;
                return cur;
            }
        }else if(root.val < key) {
            root.right = deleteNode(root.right, key);
        }else {
            root.left = deleteNode(root.left, key);
        }
        return root;
    } 
}

File: c++/tree/bst/deleteNodeInBST.cpp
#include <include/listnode.h>
/**
 * 450. Delete Node in a BST
 * 在删除中比较难想的是左右子树都存在的情况下 我们想要以右子树的最左边的node为新的node
 * 这时候可以one pass寻找+删除吗? 有点难区分第一个就是的情况(root->right->left == null)
 * 故此干脆再call一遍方法 把left most node从右边子树删掉 然后把val换到当前的node鸠占鹊巢
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(!root) return NULL;
        if(root->val > key) {
            root->left = deleteNode(root->left, key);
            return root;
        }else if(root->val < key) {
            root->right = deleteNode(root->right, key);
            return root;
        }else{
            if(!root->left && !root->right) return NULL;
            else if(!root->left && root->right) return root->right;
            else if(root->left && !root->right) return root->left;
            else{
                TreeNode* successor = findLeftMost(root->right);
                root->val = successor->val;
                root->right = deleteNode(root->right, root->val);
                return root;
            }
        }
    }
    TreeNode* findLeftMost(TreeNode* root) {
        if(!root) return NULL;
        return root->left ? findLeftMost(root->left) : root;
    }

};